Add M88 G-code to align Z-axis to full-step position

MathiasGyldenberg opened the pull request:
This pull request introduces a new G-code command `M88` that moves the Z-axis to the nearest full-step position. This enhancement addresses [Issue #2876](https://github.com/prusa3d/Prusa-Firmware/issues/2876), which proposes a method to reduce backlash and improve Z-axis reliability when steppers are disabled.

### Details

- New and unused G-code M88.
- Logic copied from power panic feature.
- Ensures the target position is within Z bounds before executing the move.
- Adds a new LCD menu item labeled `"Fullstep Z axis"` to trigger `M88` manually.
- Message string defined in `messages.cpp` and declared in `messages.h`.

### Benefits

- Reduces Z-axis skew and backlash by aligning Z motors to nearest full-step.
- Provides users with manual control via LCD menu as well as g-code command.

### Notes

- Tested on MK3S+ in both menu and end G-code.

---

### Related Issue

Fixes [#2876](https://github.com/prusa3d/Prusa-Firmware/issues/2876)

MathiasGyldenberg committed:
Add M88 command to move Z axis to full-step position

Introduces the M88 G-code command for TMC2130 drivers to move the Z axis to a full-step position. Adds corresponding message and menu item for LCD calibration when Z axis position is known.

MathiasGyldenberg committed:
Add M88 command to move Z axis to full-step position

Introduces the M88 G-code command for TMC2130 drivers to move the Z axis to a full-step position. Adds corresponding message and menu item for LCD calibration when Z axis position is known.

MathiasGyldenberg committed:
Merge remote-tracking branch 'upstream/MK3' into Fullstep-Z

MathiasGyldenberg committed:
Revert "Merge remote-tracking branch 'upstream/MK3' into Fullstep-Z"

This reverts commit 9117f184e3d7fe186cf775f0b6ee18b01375a3ec, reversing
changes made to 788d0cc63afc048fc59efa96043ff58aa87cc979.

MathiasGyldenberg committed:
Revert "Merge remote-tracking branch 'upstream/MK3' into Fullstep-Z"

This reverts commit 9117f184e3d7fe186cf775f0b6ee18b01375a3ec, reversing
changes made to 788d0cc63afc048fc59efa96043ff58aa87cc979.

3d-gussner commented:
@MathiasGyldenberg Thanks for the PR.
Looking at the https://reprap.org/wiki/G-code I would suggest to look for another gcode that might do something similar and add a parameter. This would reduce the introduction of new gcodes.

Copy the part from the power panic also uses more resources. It would probably save some bytes by making this part a new process and use it in both `uvlo` and the gcode part.

MathiasGyldenberg commented:
Thanks for your feedback @3d-gussner

I agree that adding this as a parameter to an existing G-code would simplify things. I’ve reviewed the list, but from my perspective, none of the existing commands align well with the intended purpose of aligning Z steppers to nearest fullstep.

I’m open to suggestions if there’s a better fit than introducing a dedicated command.

I’ll also refactor the logic into a reusable function so it can be called both from the G-code and power-panic.

3d-gussner commented:
`G88` has been reserved and has no function yet. As it is close to other step related gcodes. This would be a good one.

MathiasGyldenberg committed:
Refactor Z axis full-step movement to G88 command

Moved Z axis full-step positioning logic from M88 to G88 command. Extracted movement code into a new move_z_to_next_fullstep() function, updated references in power_panic and LCD menu, and removed the old M88 implementation.

MathiasGyldenberg committed:
Refactor Z axis full-step movement to G88 command

Moved Z axis full-step positioning logic from M88 to G88 command. Extracted movement code into a new move_z_to_next_fullstep() function, updated references in power_panic and LCD menu, and removed the old M88 implementation.

MathiasGyldenberg commented:
Good one with G88 - Haven't found out what it was intented to in the first place.
I've updated the commit to replace M88 with G88 as suggested. The related logic has now been refactored into a dedicated function for better clarity and reuse.

Fully tested on MK3S+

3d-gussner commented:
Please rebase your fork.
- `git remote -v` to see if you have prusa3d as upstream repository
- `git fetch --all` to get changes from the prusa3d repo
- `git rebase upstream/MK3` to update your branch `Fullstep-Z` 

there are too many changes in this PR

MathiasGyldenberg committed:
Merge branch 'Fullstep-Z' of https://github.com/MathiasGyldenberg/Prusa-Firmware into Fullstep-Z

3d-gussner commented:
Other thing would be to update the translations. See https://github.com/prusa3d/Prusa-Firmware/blob/MK3/lang/README.md

I usually make a backup of the existing `pot` and `po` files to `pot_bak` and `po_bak` before running the python scripts
- then rename the new `pot` and `po` files to `pot_new` and `po_new`
- rename the `_bak` files back to original 
- and use `Meld` to add the new messages

I know it sounds complicated BUT the translation change review is then only limited to the new messages. Here an example https://github.com/prusa3d/Prusa-Firmware/pull/4832/files#diff-14fedb8dfdce511a0bdcf251bbe0fc1c53d69fcec094186519cc6f57826b4b1b

Before a release when everything has been locked down I run the translation scripts to update everything.

But if you need some help I can do it for you and create a PR into your branch.

MathiasGyldenberg commented:
Thanks. I'm glad you mentioned if I need some help. My experience using GitHub is very limited, so any help is highly appriciated.

MathiasGyldenberg committed:
Move 'Fullstep Z axis' message and add translations

Moved MSG_FULLSTEP_Z in messages.cpp and updated internationalization files with translations for the new message across multiple languages.

MathiasGyldenberg commented:
Again thank you for your guide and help – well described.

I've updated the commit with translations, with a little translation-help from Copilot. Let me know if anything is needed from my side, and I'll be happy to assist further or make any adjustments.

MathiasGyldenberg committed:
Shortened translations for 'Fullstep Z axis' message

Shortened the German, Spanish, and French translations for the 'Fullstep Z axis' message.

MathiasGyldenberg committed:
Shortened translations for 'Fullstep Z axis' message

Shortened the translations of 'Fullstep Z axis' to 'Fullstep Z' in Hungarian, Italian, Dutch, and Romanian PO files.

MathiasGyldenberg committed:
Update G88 command documentation

Revised the comment for the G88 command
